crates/fs/private/previous.rs

```diff
@@ -0,0 +1,843 @@
+            )
+            .await
+            .unwrap();
+        let discrepancy_budget = 1_000_000;
+        let past_ratchet = root_dir.header.ratchet.clone();
+
+        let PrivateOpResult { root_dir, hamt, .. } = root_dir
+            .write(
+                &["file.txt".into()],
+                true,
+                Utc::now(),
+                b"file".to_vec(),
+                hamt,
+                store,
+                rng,
+            )
+            .await
+            .unwrap();
+
+        let PrivateOpResult { root_dir, hamt, .. } = root_dir
+            .mkdir(&["docs".into()], true, Utc::now(), hamt, store, rng)
+            .await
+            .unwrap();
+
+        let mut iterator = PrivateNodeOnPathHistory::of(
+            root_dir,
+            &past_ratchet,
+            discrepancy_budget,
+            &[],
+            true,
+            Rc::clone(&hamt),
+            store,
+        )
+        .await
+        .unwrap();
+
+        assert!(iterator.previous(store).await.unwrap().is_some());
+        assert!(iterator.previous(store).await.unwrap().is_some());
+        assert!(iterator.previous(store).await.unwrap().is_none());
+    }
+
+    /// This test will generate the following file system structure:
+    ///
+    /// (horizontal = time series, vertical = hierarchy)
+    /// ```plain
+    /// ┌────────────┐              ┌────────────┐              ┌────────────┐
+    /// │            │              │            │              │            │
+    /// │    Root    ├─────────────►│    Root    ├─────────────►│    Root    │
+    /// │            │              │            │              │            │
+    /// └────────────┘              └─────┬──────┘              └─────┬──────┘
+    ///                                   │                           │
+    ///                                   │                           │
+    ///                                   ▼                           ▼
+    ///                             ┌────────────┐              ┌────────────┐
+    ///                             │            │              │            │
+    ///                             │    Docs    ├─────────────►│    Docs    │
+    ///                             │            │              │            │
+    ///                             └─────┬──────┘              └─────┬──────┘
+    ///                                   │                           │
+    ///                                   │                           │
+    ///                                   ▼                           ▼
+    ///                             ┌────────────┐              ┌────────────┐
+    ///                             │            │              │            │
+    ///                             │  Notes.md  ├─────────────►│  Notes.md  │
+    ///                             │            │              │            │
+    ///                             └────────────┘              └────────────┘
+    /// ```
```

Comment:
Thanks for creating these ASCII diagrams.

Reply:
agreed.